### PR TITLE
Add sampling for GammaInverse Distribution

### LINKED ISSUE #7358: Use scipy.stats when sampling from sympy.stats distributions
The `sample` operation samples a random expression. 

```
>>> X = Normal('X', 0, 1)
>>> sample(100 + X)
100.325245
```

This depends on `*Distribution.sample` methods for each type of distribution (like `NormalDistribution`.)  These methods often call out to a sampling method in the `random` stdlib module.  When these methods do not exist we attempt to create one analytically by computing the CDF of the distribution.

In some cases the distribution is not found in `random` and the computation of the CDF is difficult.  When these two happen together it breaks our ability to sample.

One way to resolve this issue is to depend on the `scipy.stats` module instead of `random`, which is generally far more comprehensive.  This dependence would need to be shielded in case `scipy` wasn't present on the user's machine.

If someone wanted to fix this they should look at `sympy/stats/crv_types.py`, take a look at `GammaDistribution.sample` (works) and then try to implement the same thing for `GammaInverseDistribution.sample` (does not yet exist) using `scipy.stats.invgamma`.  Once that works there are several more!

### LINKED ISSUE #13879: sample method for GammaInverseDistribution
Currently `GammaInverseDistribution` https://github.com/sympy/sympy/blob/master/sympy/stats/crv_types.py#L1186 doesn't have a `sample` method. An attempt was made to add in PR #11673 to add `sample` method. Though the tests haven't been added.

#### What needs to be done?

* Add docstring and tests for it.
* The contributor should make sure to send a PR with ownership of original commit retained to author of PR https://github.com/sympy/sympy/pull/11673 , i.e. the submitted PR should be based on top of PR #11673.

### PR DESCRIPTION
**Pull request by RavicharanN**
Added sampling, docstring, and unit tests for GammaInverse Distribution

#### References to other Issues or PRs

Fixes  #13879. Closes #11699. 
Is a part of #7358.

#### Brief description of what is fixed or changed
Used scipy for the sampling of inverse gamma distribution as mentioned in the issue #7358. Added unit tests and modified the corresponding docstring.

#### Other comments
The docstring of the sampling method has been removed because the tests fail if scipy is not installed

### CODE REVIEW

File: sympy/stats/crv_types.py [hunk 1]
```diff
@@ -1308,6 +1314,9 @@ def GammaInverse(name, a, b):
     >>> cdf(X)(z)
     Piecewise((uppergamma(a, b/z)/gamma(a), z > 0), (0, True))
 
+    >>> Y = GammaInverse("y",1,1)
+    >>> isinstance(sample(Y),float)
```
Comment (ArighnaIITG) [inline sympy/stats/crv_types.py]: Please use PEP-8 formatting while writing code in SymPy.
The line should be `Y = GammaInverse("y", 1, 1)`
Similarly, follow this for all other cases.

**Comment by RavicharanN** [comment]
@ArighnaIITG Updated. Please check.

**Comment by RavicharanN** [comment]
Ping @asmeurer @normalhuman

File: sympy/stats/crv_types.py [hunk 2]
```diff
@@ -1260,6 +1261,13 @@ def _cdf(self, x):
         return Piecewise((uppergamma(a,b/x)/gamma(a), x > 0),
                         (S.Zero, True))
 
+    def sample(self):
+        try :
+            from scipy.stats import invgamma
+        except ImportError:
+            skip("Scipy not found")
```
Comment (normalhuman) [inline sympy/stats/crv_types.py]: This is not the place for "skip", what is being skipped when a user calls sample(X) in their console? The place for skip is in the test file, where you want to skip a test.

File: sympy/stats/crv_types.py [hunk 3]
```diff
@@ -55,6 +55,7 @@
 from sympy.stats.crv import (SingleContinuousPSpace, SingleContinuousDistribution,
         ContinuousDistributionHandmade)
 from sympy.stats.rv import _value_check
+from sympy.utilities.pytest import skip
```
Comment (normalhuman) [inline sympy/stats/crv_types.py]: This import shouldn't be here; not a test file.

File: sympy/stats/crv_types.py [hunk 4]
```diff
@@ -1260,6 +1261,13 @@ def _cdf(self, x):
         return Piecewise((uppergamma(a,b/x)/gamma(a), x > 0),
                         (S.Zero, True))
 
+    def sample(self):
+        try :
+            from scipy.stats import invgamma
```
Comment (normalhuman) [inline sympy/stats/crv_types.py]: SymPy has `import_module` in sympy.external.importtools, which checks for ImportError and many other things like incompatible versions. Use it. This try-except block should be replaced by 
```
scipy = import_module('scipy')
if scipy:
    from scipy.stats import invgamma
```
(import_module returns None if it failed.)
See https://github.com/sympy/sympy/blob/master/sympy/external/importtools.py

File: sympy/stats/tests/test_continuous_rv.py [hunk 5]
```diff
@@ -351,9 +351,10 @@ def test_gamma_inverse():
     b = Symbol("b", positive=True)
 
     X = GammaInverse("x", a, b)
+    Y = GammaInverse("y", 1, 1)
     assert density(X)(x) == x**(-a - 1)*b**a*exp(-b/x)/gamma(a)
     assert cdf(X)(x) == Piecewise((uppergamma(a, b/x)/gamma(a), x > 0), (0, True))
-
+    assert sample(Y) in Y.pspace.domain.set
```
Comment (normalhuman) [inline sympy/stats/tests/test_continuous_rv.py]: Make it a separate test (for Scipy-dependent sampling) and skip the test if import fails, as is done for example here:
https://github.com/sympy/sympy/blob/master/sympy/core/tests/test_numbers.py#L1814

File: sympy/stats/crv_types.py [hunk 6]
```diff
@@ -1287,7 +1295,7 @@ def GammaInverse(name, a, b):
     Examples
     ========
 
-    >>> from sympy.stats import GammaInverse, density, cdf, E, variance
+    >>> from sympy.stats import GammaInverse, density, cdf, E, variance, sample
```
Comment (normalhuman) [inline sympy/stats/crv_types.py]: Why is sample imported here? It's not used in the docstring.

**Comment by RavicharanN** [comment]
@normalhuman Updated the PR. Please review

File: sympy/stats/crv_types.py [hunk 7]
```diff
@@ -1260,6 +1261,11 @@ def _cdf(self, x):
         return Piecewise((uppergamma(a,b/x)/gamma(a), x > 0),
                         (S.Zero, True))
 
+    def sample(self):
+        scipy = import_module('scipy')
+        if scipy:
```
Comment (normalhuman) [inline sympy/stats/crv_types.py]: And if not, should an error be raised? Like `NotImplementedError('Sampling the inverse Gamma distribution requires SciPy')`? As is, the method returns None.

File: sympy/stats/tests/test_continuous_rv.py [hunk 8]
```diff
@@ -354,6 +355,12 @@ def test_gamma_inverse():
     assert density(X)(x) == x**(-a - 1)*b**a*exp(-b/x)/gamma(a)
     assert cdf(X)(x) == Piecewise((uppergamma(a, b/x)/gamma(a), x > 0), (0, True))
 
+def test_sampling_gamma_inverse():
+    scipy = import_module('scipy')
+    if not scipy:
+        skip('scipy not installed. Abort tests fro sampling of gamma inverse.')
```
Comment (normalhuman) [inline sympy/stats/tests/test_continuous_rv.py]: fro -> for
scipy -> SciPy

**Comment by RavicharanN** [comment]
ping @normalhuman

**Comment by normalhuman** [comment]
I edited PR description, because it does not really fix all of #7358. Also, as originally written, it would not automatically close #13879. The word "fixes" or "closes" must be repeated before each issue or PR that should be closed. See https://github.com/sympy/sympy/wiki/Issue-PR-Autoclosing-syntax

**Pull request merged**